### ISSUE TITLE
NVR support

### ISSUE THREAD

**Issue by kukulich** [rroller/dahua#9]
Is it possible to support NVR? It would be nice to add one device (NVR) and all info about cameras would be detected automatically.

It should also solve a problem when cameras cannot be added because they are connected via NVR and NVR use specific custom IP addresses for them - and the IP addresses are not usable directly.

**Comment by rroller** [comment]
I think that's doable. But I don't have an nvr and I'd need one to build out the functionality.

**Comment by kukulich** [comment]
I can help with debugging and testing. Is there a list of API somewhere?

**Comment by rroller** [comment]
The problem is that I need an actual NVR so I can do edit-test cycles locally to get all the APIs correct. Without that I have no idea if my code would work or even where to start. I have to play around with the APIs to understand them and for that I need an NVR unit.

**Comment by kukulich** [comment]
I have already tested some API:

`/cgi-bin/magicBox.cgi?action=getSystemInfo`
```
deviceType=31
processor=ST7108
serialNumber=4F07335PAZEEA9B
updateSerial=NVR2X-4KS3
```

`/cgi-bin/magicBox.cgi?action=getSoftwareVersion`
```
version=4.001.0000005.0,build:2020-11-17
```

`/cgi-bin/magicBox.cgi?action=getMachineName`
```
name=NVR
```

`/cgi-bin/magicBox.cgi?action=getVendor`
```
vendor=Dahua
```

But I don't know what can I test next... If you can guide me, I may be able to send a PR...

**Comment by kukulich** [comment]
`/cgi-bin/magicBox.cgi?action=getDeviceType`
```
type=DHI-NVR2108-8P-4KS2
```

`/cgi-bin/magicBox.cgi?action=getDeviceClass`
```
class=NVR
```

**Comment by kukulich** [comment]
This looks like a way to detect number of channels:

`/cgi-bin/configManager.cgi?action=getConfig&name=VideoColor`

```
table.VideoColor[0][0].Brightness=50
table.VideoColor[0][0].ChromaSuppress=50
table.VideoColor[0][0].Contrast=50
table.VideoColor[0][0].Gamma=50
table.VideoColor[0][0].Hue=50
table.VideoColor[0][0].Saturation=50
table.VideoColor[0][0].Style=Standard
table.VideoColor[0][0].TimeSection=1 00:00:00-24:00:00
table.VideoColor[0][1].Brightness=50
table.VideoColor[0][1].ChromaSuppress=50
table.VideoColor[0][1].Contrast=50
table.VideoColor[0][1].Gamma=50
table.VideoColor[0][1].Hue=50
table.VideoColor[0][1].Saturation=50
table.VideoColor[0][1].Style=Standard
table.VideoColor[0][1].TimeSection=0 00:00:00-24:00:00
table.VideoColor[0][2].Brightness=50
table.VideoColor[0][2].ChromaSuppress=50
table.VideoColor[0][2].Contrast=50
table.VideoColor[0][2].Gamma=50
table.VideoColor[0][2].Hue=50
table.VideoColor[0][2].Saturation=50
table.VideoColor[0][2].Style=Standard
table.VideoColor[0][2].TimeSection=0 00:00:00-24:00:00
table.VideoColor[1][0].Brightness=50
table.VideoColor[1][0].ChromaSuppress=50
table.VideoColor[1][0].Contrast=50
table.VideoColor[1][0].Gamma=50
table.VideoColor[1][0].Hue=50
table.VideoColor[1][0].Saturation=50
table.VideoColor[1][0].Style=Standard
table.VideoColor[1][0].TimeSection=1 00:00:00-24:00:00
table.VideoColor[1][1].Brightness=50
table.VideoColor[1][1].ChromaSuppress=50
table.VideoColor[1][1].Contrast=50
table.VideoColor[1][1].Gamma=50
table.VideoColor[1][1].Hue=50
table.VideoColor[1][1].Saturation=50
table.VideoColor[1][1].Style=Standard
table.VideoColor[1][1].TimeSection=0 00:00:00-24:00:00
table.VideoColor[1][2].Brightness=50
table.VideoColor[1][2].ChromaSuppress=50
table.VideoColor[1][2].Contrast=50
table.VideoColor[1][2].Gamma=50
table.VideoColor[1][2].Hue=50
table.VideoColor[1][2].Saturation=50
table.VideoColor[1][2].Style=Standard
table.VideoColor[1][2].TimeSection=0 00:00:00-24:00:00
table.VideoColor[2][0].Brightness=50
table.VideoColor[2][0].ChromaSuppress=50
table.VideoColor[2][0].Contrast=50
table.VideoColor[2][0].Gamma=50
table.VideoColor[2][0].Hue=50
table.VideoColor[2][0].Saturation=50
table.VideoColor[2][0].Style=Standard
table.VideoColor[2][0].TimeSection=1 00:00:00-24:00:00
table.VideoColor[2][1].Brightness=50
table.VideoColor[2][1].ChromaSuppress=50
table.VideoColor[2][1].Contrast=50
table.VideoColor[2][1].Gamma=50
table.VideoColor[2][1].Hue=50
table.VideoColor[2][1].Saturation=50
table.VideoColor[2][1].Style=Standard
table.VideoColor[2][1].TimeSection=0 00:00:00-24:00:00
table.VideoColor[2][2].Brightness=50
table.VideoColor[2][2].ChromaSuppress=50
table.VideoColor[2][2].Contrast=50
table.VideoColor[2][2].Gamma=50
table.VideoColor[2][2].Hue=50
table.VideoColor[2][2].Saturation=50
table.VideoColor[2][2].Style=Standard
table.VideoColor[2][2].TimeSection=0 00:00:00-24:00:00
table.VideoColor[3][0].Brightness=50
table.VideoColor[3][0].ChromaSuppress=50
table.VideoColor[3][0].Contrast=50
table.VideoColor[3][0].Gamma=50
table.VideoColor[3][0].Hue=50
table.VideoColor[3][0].Saturation=50
table.VideoColor[3][0].Style=Standard
table.VideoColor[3][0].TimeSection=1 00:00:00-24:00:00
table.VideoColor[3][1].Brightness=50
table.VideoColor[3][1].ChromaSuppress=50
table.VideoColor[3][1].Contrast=50
table.VideoColor[3][1].Gamma=50
table.VideoColor[3][1].Hue=50
table.VideoColor[3][1].Saturation=50
table.VideoColor[3][1].Style=Standard
table.VideoColor[3][1].TimeSection=0 00:00:00-24:00:00
table.VideoColor[3][2].Brightness=50
table.VideoColor[3][2].ChromaSuppress=50
table.VideoColor[3][2].Contrast=50
table.VideoColor[3][2].Gamma=50
table.VideoColor[3][2].Hue=50
table.VideoColor[3][2].Saturation=50
table.VideoColor[3][2].Style=Standard
table.VideoColor[3][2].TimeSection=0 00:00:00-24:00:00
table.VideoColor[4][0].Brightness=50
table.VideoColor[4][0].ChromaSuppress=50
table.VideoColor[4][0].Contrast=50
table.VideoColor[4][0].Gamma=50
table.VideoColor[4][0].Hue=50
table.VideoColor[4][0].Saturation=50
table.VideoColor[4][0].Style=Standard
table.VideoColor[4][0].TimeSection=1 00:00:00-24:00:00
table.VideoColor[4][1].Brightness=50
table.VideoColor[4][1].ChromaSuppress=50
table.VideoColor[4][1].Contrast=50
table.VideoColor[4][1].Gamma=50
table.VideoColor[4][1].Hue=50
table.VideoColor[4][1].Saturation=50
table.VideoColor[4][1].Style=Standard
table.VideoColor[4][1].TimeSection=0 00:00:00-24:00:00
table.VideoColor[4][2].Brightness=50
table.VideoColor[4][2].ChromaSuppress=50
table.VideoColor[4][2].Contrast=50
table.VideoColor[4][2].Gamma=50
table.VideoColor[4][2].Hue=50
table.VideoColor[4][2].Saturation=50
table.VideoColor[4][2].Style=Standard
table.VideoColor[4][2].TimeSection=0 00:00:00-24:00:00
table.VideoColor[5][0].Brightness=50
table.VideoColor[5][0].ChromaSuppress=50
table.VideoColor[5][0].Contrast=50
table.VideoColor[5][0].Gamma=50
table.VideoColor[5][0].Hue=50
table.VideoColor[5][0].Saturation=50
table.VideoColor[5][0].Style=Standard
table.VideoColor[5][0].TimeSection=1 00:00:00-24:00:00
table.VideoColor[5][1].Brightness=50
table.VideoColor[5][1].ChromaSuppress=50
table.VideoColor[5][1].Contrast=50
table.VideoColor[5][1].Gamma=50
table.VideoColor[5][1].Hue=50
table.VideoColor[5][1].Saturation=50
table.VideoColor[5][1].Style=Standard
table.VideoColor[5][1].TimeSection=0 00:00:00-24:00:00
table.VideoColor[5][2].Brightness=50
table.VideoColor[5][2].ChromaSuppress=50
table.VideoColor[5][2].Contrast=50
table.VideoColor[5][2].Gamma=50
table.VideoColor[5][2].Hue=50
table.VideoColor[5][2].Saturation=50
table.VideoColor[5][2].Style=Standard
table.VideoColor[5][2].TimeSection=0 00:00:00-24:00:00
table.VideoColor[6][0].Brightness=50
table.VideoColor[6][0].ChromaSuppress=50
table.VideoColor[6][0].Contrast=50
table.VideoColor[6][0].Gamma=50
table.VideoColor[6][0].Hue=50
table.VideoColor[6][0].Saturation=50
table.VideoColor[6][0].Style=Standard
table.VideoColor[6][0].TimeSection=1 00:00:00-24:00:00
table.VideoColor[6][1].Brightness=50
table.VideoColor[6][1].ChromaSuppress=50
table.VideoColor[6][1].Contrast=50
table.VideoColor[6][1].Gamma=50
table.VideoColor[6][1].Hue=50
table.VideoColor[6][1].Saturation=50
table.VideoColor[6][1].Style=Standard
table.VideoColor[6][1].TimeSection=0 00:00:00-24:00:00
table.VideoColor[6][2].Brightness=50
table.VideoColor[6][2].ChromaSuppress=50
table.VideoColor[6][2].Contrast=50
table.VideoColor[6][2].Gamma=50
table.VideoColor[6][2].Hue=50
table.VideoColor[6][2].Saturation=50
table.VideoColor[6][2].Style=Standard
table.VideoColor[6][2].TimeSection=0 00:00:00-24:00:00
table.VideoColor[7][0].Brightness=50
table.VideoColor[7][0].Contrast=50
table.VideoColor[7][0].Hue=50
table.VideoColor[7][0].Saturation=50
table.VideoColor[7][0].TimeSection=1 00:00:00-24:00:00
table.VideoColor[7][1].Brightness=50
table.VideoColor[7][1].Contrast=50
table.VideoColor[7][1].Hue=50
table.VideoColor[7][1].Saturation=50
table.VideoColor[7][1].TimeSection=0 00:00:00-24:00:00
```

**Comment by alanpilz** [comment]
I’m also interested in the NVR integration. I have a Dahua 4108 NVR, installed this great integration and I see “main” camera added and I have a preview of one of the cameras. But only one camera is visible. I simply added the IP of the dahua NVR and it detected 1 Device with 5 entities.

**Comment by kukulich** [comment]
I'll try to prepare PR.

**Comment by frapersan** [comment]
I have a nvr (dahua hcvr5108hs-s3) when I configure the integration it works, but on the integrations page it appears as:

![Captura](https://user-images.githubusercontent.com/16254354/122184109-30899b80-ce8c-11eb-8e48-43362ccd7626.PNG)

no warning appears in the logs. 

I can help, I am fully available

**Comment by chali272** [comment]
im also using dahua nvr and like to see it supported, when adding the nvr today im getting one camera.

**Comment by kukulich** [comment]
I did not find any API to detect number of cameras... However it may be possible to detect if the device is NVR (via `getDeviceClass`) so the best option would be to add second config_flow step - user can specify the number of cameras/channels by itself.

**Comment by Mikefila** [comment]
Okay So I was going through the code and in camera.py I saw this

> For now we'll only support 1 channel. I don't have any cams where I can test a second channel.
> I'm not really sure what channel 2 means anyways, it doesn't seem to be the substream.
> CHANNEL = 1

Channel refers to dvr/xvr/nvr channels. `subtype` denotes main(0)/sub stream(1)  Rtsp streams are in the form of 
`rtsp://<username>:<password>@<ip>:<port>/cam/realmonitor?channel=1&subtype=0`

Channels are offset by one. Channel 0 = Camera 1; 1=2; 2=3 etc.

I use this right now and I able to get all events from an xvr running 11 analog cameras.

https://github.com/algirdasc/appdaemon-apps/blob/master/dahua_mqtt.py

**Comment by Mikefila** [comment]
Probably the most important part, the only thing I had to modify was a setting. Instead of defining a camera in the form of `camera/1` I removed the number and just used `camera`

**Comment by chali272** [comment]
> Okay So I was going through the code and in camera.py I saw this
> 
> > For now we'll only support 1 channel. I don't have any cams where I can test a second channel.
> > I'm not really sure what channel 2 means anyways, it doesn't seem to be the substream.
> > CHANNEL = 1
> 
> Channel refers to dvr/xvr/nvr channels. `subtype` denotes main(0)/sub stream(1) Rtsp streams are in the form of
> `rtsp://<username>:<password>@<ip>:<port>/cam/realmonitor?channel=1&subtype=0`
> 
> Channels are offset by one. Channel 0 = Camera 1; 1=2; 2=3 etc.
> 
> I use this right now and I am able to get all events from an xvr running 11 analog cameras.
> 
> https://github.com/algirdasc/appdaemon-apps/blob/master/dahua_mqtt.py

noob questone, what file is this replacing in the Dahua folder/whats should your file be named as? (i cant see any file named dahua_mqtt.py)

**Comment by Mikefila** [comment]
I'm sorry if I was unclear, I did not change anything on this integration nor do I have it working. 

The link I posted is a different integration all together. It works through appdaemon. I only referenced so the maintainer of this integration has a working example of a script that will log on to a nvr/dvr/xvr. Perhaps it maybe helpful. 

Follow this post for details of the other integration, if you have further questions please post them to that thread. I'm subscribed and will answer there as not to derail this thread.

https://community.home-assistant.io/t/dahua-ipc-to-mqtt-app/93327/174?u=mikefila

**Comment by fokcuk** [comment]
what is this section undr __init.py__?
    def __init__(self, hass: HomeAssistant, events: list, address: str, port: int, rtsp_port: int, username: str,
                 password: str) -> None:
        """Initialize."""
        self.client: DahuaClient = DahuaClient(username, password, address, port, rtsp_port,
                                               async_get_clientsession(hass))
        self.dahua_event: DahuaEventThread
        self.platforms = []
        self.initialized = False
        self.model = ""
        self.machine_name = ""
        self.connected = None
        **self.channels = {"1": "1"}**
        self.events: list = events

If we use NVR address, it works and events are registered, but only for channel 1. Also there is no way to modify entity name, so even if we modify the channel in the code, it will link it to the same one.

**Comment by rroller** [comment]
Yeah I have a hard coded channel coded through the code. I can make that an option when configuring the integration to allow you to pick the channel. I think that will work.

**Comment by fokcuk** [comment]
> Yeah I have a hard coded channel coded through the code. I can make that an option when configuring the integration to allow you to pick the channel. I think that will work.

And make am option for entity name pick, otherwise it assigns the same name to the same NVR when adding and no option to change it.
Which file picks the channel for events? I saw in one there is 1:1 for a channel, but don't know it its a start:end (like 1:5) or if you want a particular channel you have to select 2:2, 3:3 etc

**Comment by rroller** [comment]
I'm close to completing a change that allows one to set the name during configuration time. This will be the building blocks needed to support channels.

It's basically the same configuration form now, but after you click submit, it'll open a new form with the name of the camera and lets you edit it.

**Comment by fokcuk** [comment]
great! thank you. 
How does it get the motion events? I thought I changed the channel info in the code but I think it reads any motion event and NVR will have different events for different channels/cameras

**Comment by fokcuk** [comment]
Platform dahua does not generate unique IDs. ID 3CPAZ5EF403317U_motion_detection already exists - ignoring switch.nvr_motion_detection
Platform dahua does not generate unique IDs. ID 3CPAZ5EF403317U_disarming already exists - ignoring switch.nvr_disarming

Can you please add the sensor unique IDs as well?

**Comment by haid45** [comment]
> I'm close to completing a change that allows one to set the name during configuration time. This will be the building blocks needed to support channels.
> 
> It's basically the same configuration form now, but after you click submit, it'll open a new form with the name of the camera and lets you edit it.

@rroller Ill be eager to test this when you have finished.

**Comment by rroller** [comment]
Thanks. I'll make this the next thing I work on. I need to un-hard-code all the channels I currently have hard coded

**Comment by haid45** [comment]
Amazing, I'm readily available to test this out for you. I currently have an NVR installed with 4 cameras (channels). The integration is able to pick up just the main channel at the moment.

**Comment by rroller** [comment]
@haid45 when you added your camera, for the IP address, do you use your NVR IP Address? I assume you don't use the IP of the camera? So you'd be adding 4 camera's all with the IP of your NVR?

**Comment by haid45** [comment]
@rroller, yes I used the IP address of the NVR as I wasn't able to get the individual ip's of the cameras. Well actually, I was but they seemed to be on a local network within the NVR because the ip's where along the lines of 10.0.0.1.
Adding the 4 cameras with IP of the NVR is definitely the end game here.

**Comment by rroller** [comment]
Awesome, this is totally doable. I'm working on the changes now.

**Comment by haid45** [comment]
How exciting! looking forward to trying it out 🥳

**Comment by rroller** [comment]
I just released 0.9.0 with early support for NVR's. I don't have an NVR so I can't test it... so this will most definitely have bugs :) 

If you all could please test it and let me know if you run into any issues and I'll try to fix them ASAP.

https://github.com/rroller/dahua/releases/tag/0.9.0

**Comment by rroller** [comment]
And now 0.9.1 is released for a small bug fix https://github.com/rroller/dahua/releases/tag/0.9.1

**Comment by rroller** [comment]
One area that I know needs work is the event stream. I'll try to work on that soon. It's probably only handling events from camera 0

**Comment by rroller** [comment]
And I just realized another change I need to make for this to work. The unique id used for the entities is the serial number. All cams added will get the nvr serial, so they’ll collide. I’ll need to add in the channel to the unique id.

**Comment by haid45** [comment]
@rroller, thank you for that. It's great!
I have just tested it on my setup and the first hurdle is that by adding the NVR it does not add all the cameras that are associated with the NVR. I tried adding multiple integrations, one for each channel but received this error in the process `Only a single instance of a device is allowed.`

**Comment by rroller** [comment]
@haid45 thsts the issue I mentioned above. I'll try to fix that today.

**Comment by rroller** [comment]
OK 0.9.2 is out. Try that one.  This will hopefully let you add the additional cameras and from there we can debug any issues. Known issues that I know about are events. Once we get the additional cams added I'll ask if one of you can provide me event logs so I can then fix that. Sorry for so many iterations here, hard without having an NVR :D 
https://github.com/rroller/dahua/releases/tag/0.9.2

**Comment by mrbarbz** [comment]
I updated to version 0.9.2, I still have the same issue in #50 with your latest update.

I also can't connect a single camera, I think its because the NVR has an internal PoE switch and assigns different Ip addresses from the NVR
Example NVR IP address 10.0.0.5,  Camera on the NVR IP 10.2.2.9

**Comment by mrbarbz** [comment]
> I updated to version 0.9.2, I still have the same issue in #50 with your latest update.
> 
> I also can't connect a single camera, I think its because the NVR has an internal PoE switch and assigns different Ip addresses from the NVR
> Example NVR IP address 10.0.0.5, Camera on the NVR IP 10.2.2.9

It's working now. Thanks heaps
When I was trying to sort out the issue I found that there was new Dahua NVR system Firmware. I upgraded to version 4 and like magic no more error.
I will play around and see if there are any issues I come across.
Thanks again

**Comment by rroller** [comment]
Great! Ill note the firmware update in the readme. Also please let me know if you can grab some of the even json. You'll need to enable debug logging, there's an example in the readme.

**Comment by mrbarbz** [comment]
I have debugged logging already enabled but where do I get it from?

**Comment by mrbarbz** [comment]
After some testing, this is what I have experienced 

When I updated the NVR firmware and tested it, I had it on my desk. 
I tested with 1 camera, which has its own POE injector and is connected to my network router.
With only 1 camera connected and I could add the NVR using its Ip address.
I put back the NVR in the carboard and connected the rest of the cameras.
I removed the integration, restarted HA and re-added it to that it can pick up all the cameras, every time I put in my details and select submit it would clear my inputs giving me the "username/password/Ip address" wrong error.
These are the following IPs
NVR 10.0.0.100
cam1 10.0.0.101
cam2 10.0.0102
cam3 10.0.0.103
cam4 10.0.0.104


cam4 is the only camera not connected to the NVR's internal PoE switch, it is connected to my network with PoE injector

Every time I try to add cam1, cam2, cam3, and NVR with the relevant IP it would get the username/password/IP address error
cam4 is the only camera I could connect to.

All cameras have also been updated to Version 2.8

I don't know if the issue is within your integration or my internal switch settings.

**Comment by haid45** [comment]
In my case, I used the ip address of the nvr and just added multiple integrations, one for each channel. So far it works great but I have not utilised any of the sensors or switches, just the cameras.

**Comment by fokcuk** [comment]
camera integration you can do with a simple rtsp stream. 
Events per channel are what I need from the integration. At the moment I am still running the older version of the integration that I modified py code to point to a particular camera for video. Channel events were not available then (don't know about now). Any event from NVR (events from any channel) would trigger the entity change for things like cross line etc. I went around it by having a particular alarm that is only specific to the channel I want

**Comment by fokcuk** [comment]
@rroller - check your code in camera.py (and maybe others).
You have **channel + 1** in a number of places in the code. It may work if channel is 0 (zero), but for multi channel NVR it breaks things. I only have 5 channels, but it tries to get stream from channel 6

**Comment by mrbarbz** [comment]
An update to my issue
looks like my issue is not related to your integration.

This is what I have found 
The NVR will have DHCP and will give a different subnet to its cameras .. this is where the problem is ...
So the issue is with the inbuilt Poe Switch on the NVR having its own DHCP. 

I couldn't find a way to disable it and I don't think there is an option. I am going to purchase a separate PoE switch, connect cameras and NVR to it and test it again.

**Comment by mrbarbz** [comment]
ok purchased PoE Switch

Integration wouldn't connect to NVR, I kept getting either username/password/IP address incorrect

Integration worked with the cameras via their IP address connected to the same network as HA via PoE switch but video stream only worked when I left the channel at 0 at setup.

I've found that I could only connect to the cameras when not using internal/inbuilt switch

 Below is the log when I try to connect to the NVR if that is any help to you. 


This error originated from a custom integration.

Logger: custom_components.dahua
Source: custom_components/dahua/digest.py:46
Integration: Dahua (documentation, issues)
First occurred: 8:32:24 PM (1 occurrences)
Last logged: 8:32:24 PM

Could not connect to Dahua device
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 969, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1050, in create_connection
    transport, protocol = await self._create_connection_transport(
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1080, in _create_connection_transport
    await waiter
  File "/usr/local/lib/python3.8/asyncio/sslproto.py", line 529, in data_received
    ssldata, appdata = self._sslpipe.feed_ssldata(data)
  File "/usr/local/lib/python3.8/asyncio/sslproto.py", line 189, in feed_ssldata
    self._sslobj.do_handshake()
  File "/usr/local/lib/python3.8/ssl.py", line 944, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1125)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/dahua/config_flow.py", line 201, in _test_credentials
    data = await client2.get_machine_name()
  File "/config/custom_components/dahua/client.py", line 95, in get_machine_name
    return await self.get(url)
  File "/config/custom_components/dahua/client.py", line 565, in get
    raise exception
  File "/config/custom_components/dahua/client.py", line 550, in get
    response = await auth.request("GET", url)
  File "/config/custom_components/dahua/digest.py", line 46, in request
    response = await self.session.request(method, url, headers=headers, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/client.py", line 520, in _request
    conn = await self._connector.connect(
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 535, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 892, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 1051, in _create_direct_connection
    raise last_exc
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 1020, in _create_direct_connection
    transp, proto = await self._wrap_create_connection(
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 971, in _wrap_create_connection
    raise ClientConnectorCertificateError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorCertificateError: Cannot connect to host 10.0.0.100:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1125)')]

**Comment by mrbarbz** [comment]
> ok purchased PoE Switch
> 
> Integration wouldn't connect to NVR, I kept getting either username/password/IP address incorrect
> 
> Integration worked with the cameras via their IP address connected to the same network as HA via PoE switch but video stream only worked when I left the channel at 0 at setup.
> 
> I've found that I could only connect to the cameras when not using internal/inbuilt switch
> 
> Below is the log when I try to connect to the NVR if that is any help to you.
> 
> This error originated from a custom integration.
> 
> Logger: custom_components.dahua
> Source: custom_components/dahua/digest.py:46
> Integration: Dahua (documentation, issues)
> First occurred: 8:32:24 PM (1 occurrences)
> Last logged: 8:32:24 PM
> 
> Could not connect to Dahua device
> Traceback (most recent call last):
> File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 969, in _wrap_create_connection
> return await self._loop.create_connection(*args, **kwargs) # type: ignore # noqa
> File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1050, in create_connection
> transport, protocol = await self._create_connection_transport(
> File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1080, in _create_connection_transport
> await waiter
> File "/usr/local/lib/python3.8/asyncio/sslproto.py", line 529, in data_received
> ssldata, appdata = self._sslpipe.feed_ssldata(data)
> File "/usr/local/lib/python3.8/asyncio/sslproto.py", line 189, in feed_ssldata
> self._sslobj.do_handshake()
> File "/usr/local/lib/python3.8/ssl.py", line 944, in do_handshake
> self._sslobj.do_handshake()
> ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1125)
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
> File "/config/custom_components/dahua/config_flow.py", line 201, in _test_credentials
> data = await client2.get_machine_name()
> File "/config/custom_components/dahua/client.py", line 95, in get_machine_name
> return await self.get(url)
> File "/config/custom_components/dahua/client.py", line 565, in get
> raise exception
> File "/config/custom_components/dahua/client.py", line 550, in get
> response = await auth.request("GET", url)
> File "/config/custom_components/dahua/digest.py", line 46, in request
> response = await self.session.request(method, url, headers=headers, **kwargs)
> File "/usr/local/lib/python3.8/site-packages/aiohttp/client.py", line 520, in _request
> conn = await self._connector.connect(
> File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 535, in connect
> proto = await self._create_connection(req, traces, timeout)
> File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 892, in _create_connection
> _, proto = await self._create_direct_connection(req, traces, timeout)
> File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 1051, in _create_direct_connection
> raise last_exc
> File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 1020, in _create_direct_connection
> transp, proto = await self._wrap_create_connection(
> File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 971, in _wrap_create_connection
> raise ClientConnectorCertificateError(req.connection_key, exc) from exc
> aiohttp.client_exceptions.ClientConnectorCertificateError: Cannot connect to host 10.0.0.100:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1125)')]


Also, got this error when trying to connect NVR

This error originated from a custom integration.

Logger: custom_components.dahua
Source: custom_components/dahua/config_flow.py:197
Integration: Dahua (documentation, issues)
First occurred: 8:32:24 PM (1 occurrences)
Last logged: 8:32:24 PM

Could not connect to Dahua device via the RPC2 API, falling back to cgi

**Comment by fokcuk** [comment]
try to disable SSL and try again. ONVIF on my Dahua NVR has some sort of issue with SSL

**Comment by mrbarbz** [comment]
> try to disable SSL and try again. ONVIF on my Dahua NVR has some sort of issue with SSL

You mean disable on the NVR? If so how? I couldn't find it on the NVR

**Comment by fokcuk** [comment]
on mine its under Settings>Security>CA Certificates

**Comment by rroller** [comment]
I've added SSL support in the latest version on this integration. Do you mind trying again? It requires that your port is 443.

**Comment by rroller** [comment]
@fokcuk are you still having issues with events? If so I'd like to help get that solved.

Can you enable debug logging and paste some of the events from the logs?

You can enable debug logging like this...

```yaml
logger:
  default: info
  logs:
    custom_components.dahua: debug
```

**Comment by fokcuk** [comment]
> @fokcuk are you still having issues with events? If so I'd like to help get that solved.
> 
> Can you enable debug logging and paste some of the events from the logs?
> 
> You can enable debug logging like this...
> 
> ```yaml
> logger:
>   default: info
>   logs:
>     custom_components.dahua: debug
> ```

You still have not fixed Channel+1 issue. I picked channel 5 but its trying to read channel 6

**Comment by fokcuk** [comment]
events are not fixed. Still get the events from other channels
The alarm event in NVR does contain channel info. If you can tell me how to check it raw, I can paste the output string

**Comment by kordekm** [comment]
I was able to connect via SSL and port 443. Unfortunately, switches such as IR do not work. I switch it on and after few secounds switch goes back to off . Nothing happends with camera. 
I also do not see the switch to turn on the lights and sirens for the TIOC model (IPC-HDW3549HP-AS-PV). I saw my camera on supported models list.
 Am I doing somthing wrong?

**Comment by fokcuk** [comment]
Here is an example of events from other channels:
`2021-07-20 18:36:09 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'NVR_Ch5', 'Code': 'CrossLineDetection', 'action': 'Start', 'index': '0', 'data': {'Class': 'Normal', 'CountInGroup': 1, 'DetectLine': [[346, 3549], [1110, 4205], [346, 5250], [1511, 5688], [291, 6903], [1638, 8069]], 'Direction': 'LeftToRight', 'EventSeq': 2, 'FrameSequence': 163535, 'GroupID': 2, 'IndexInGroup': 0, 'LocaleTime': '20-07-2021 18:36:09', 'Mark': 0, 'Name': 'Rule5', 'Object': {'Action': 'Appear', 'BoundingBox': [256, 4912, 2464, 6848], 'Center': [1360, 5880], 'Confidence': 0, 'FrameSequence': 0, 'LowerBodyColor': [0, 0, 0, 0], 'MainColor': [0, 0, 0, 0], 'ObjectID': 1618, 'ObjectType': 'Human', 'RelativeID': 0, 'Source': 0.0, 'Speed': 0, 'SpeedTypeInternal': 0}, 'PTS': 43243821690.0, 'RuleId': 5, 'Sequence': 0, 'Source': 62300200.0, 'Track': None, 'UTC': 1626777369.0, 'UTCMS': 62}, 'DeviceName': 'NVR_Ch5'}
2021-07-20 18:36:11 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'NVR_Ch5', 'Code': 'VideoMotion', 'action': 'Start', 'index': '1', 'channel': '1', 'DeviceName': 'NVR_Ch5'}
2021-07-20 18:36:12 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'NVR_Ch5', 'Code': 'CrossLineDetection', 'action': 'Stop', 'index': '0', 'data': {'Class': 'Normal', 'CountInGroup': 1, 'DetectLine': [[346, 3549], [1110, 4205], [346, 5250], [1511, 5688], [291, 6903], [1638, 8069]], 'Direction': 'LeftToRight', 'EventSeq': 2, 'FrameSequence': 163535, 'GroupID': 2, 'IndexInGroup': 0, 'LocaleTime': '20-07-2021 18:36:11', 'Mark': 0, 'Name': 'Rule5', 'Object': {'Action': 'Appear', 'BoundingBox': [256, 4912, 2464, 6848], 'Center': [1360, 5880], 'Confidence': 0, 'FrameSequence': 0, 'LowerBodyColor': [0, 0, 0, 0], 'MainColor': [0, 0, 0, 0], 'ObjectID': 1618, 'ObjectType': 'Human', 'RelativeID': 0, 'Source': 0.0, 'Speed': 0, 'SpeedTypeInternal': 0}, 'PTS': 43243821690.0, 'RuleId': 5, 'Sequence': 0, 'Source': 62300200.0, 'Track': None, 'UTC': 1626777371.0, 'UTCMS': 62}, 'DeviceName': 'NVR_Ch5'}
2021-07-20 18:36:13 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'NVR_Ch5', 'Code': 'CrossLineDetection', 'action': 'Start', 'index': '0', 'data': {'Class': 'Normal', 'CountInGroup': 1, 'DetectLine': [[673, 3500], [1620, 2455], [6553, 2090], [7135, 3743]], 'Direction': 'RightToLeft', 'EventSeq': 19, 'FrameSequence': 163570, 'GroupID': 19, 'IndexInGroup': 0, 'LocaleTime': '20-07-2021 18:36:12', 'Mark': 0, 'Name': 'Rule1', 'Object': {'Action': 'Appear', 'BoundingBox': [40, 152, 3368, 3736], 'Center': [1704, 1944], 'Confidence': 0, 'FrameSequence': 0, 'LowerBodyColor': [0, 0, 0, 0], 'MainColor': [0, 0, 0, 0], 'ObjectID': 1621, 'ObjectType': 'Human', 'RelativeID': 0, 'Source': 0.0, 'Speed': 0, 'SpeedTypeInternal': 0}, 'PTS': 43243825190.0, 'RuleId': 1, 'Sequence': 0, 'Source': 62300200.0, 'Track': None, 'UTC': 1626777372.0, 'UTCMS': 557}, 'DeviceName': 'NVR_Ch5'}
2021-07-20 18:36:15 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'NVR_Ch5', 'Code': 'CrossLineDetection', 'action': 'Start', 'index': '0', 'data': {'Class': 'Normal', 'CountInGroup': 1, 'DetectLine': [[582, 1750], [673, 0]], 'Direction': 'RightToLeft', 'EventSeq': 22, 'FrameSequence': 163590, 'GroupID': 22, 'IndexInGroup': 0, 'LocaleTime': '20-07-2021 18:36:14', 'Mark': 0, 'Name': 'Rule2', 'Object': {'Action': 'Appear', 'BoundingBox': [464, 560, 720, 768], 'Center': [592, 664], 'Confidence': 0, 'FrameSequence': 0, 'LowerBodyColor': [0, 0, 0, 0], 'MainColor': [0, 0, 0, 0], 'ObjectID': 1621, 'ObjectType': 'Human', 'RelativeID': 0, 'Source': 0.0, 'Speed': 0, 'SpeedTypeInternal': 0}, 'PTS': 43243827190.0, 'RuleId': 2, 'Sequence': 0, 'Source': 62300200.0, 'Track': None, 'UTC': 1626777374.0, 'UTCMS': 563}, 'DeviceName': 'NVR_Ch5'}
2021-07-20 18:36:17 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'NVR_Ch5', 'Code': 'CrossLineDetection', 'action': 'Stop', 'index': '0', 'data': {'Class': 'Normal', 'CountInGroup': 1, 'DetectLine': [[582, 1750], [673, 0]], 'Direction': 'RightToLeft', 'EventSeq': 22, 'FrameSequence': 163590, 'GroupID': 22, 'IndexInGroup': 0, 'LocaleTime': '20-07-2021 18:36:16', 'Mark': 0, 'Name': 'Rule2', 'Object': {'Action': 'Appear', 'BoundingBox': [464, 560, 720, 768], 'Center': [592, 664], 'Confidence': 0, 'FrameSequence': 0, 'LowerBodyColor': [0, 0, 0, 0], 'MainColor': [0, 0, 0, 0], 'ObjectID': 1621, 'ObjectType': 'Human', 'RelativeID': 0, 'Source': 0.0, 'Speed': 0, 'SpeedTypeInternal': 0}, 'PTS': 43243827190.0, 'RuleId': 2, 'Sequence': 0, 'Source': 62300200.0, 'Track': None, 'UTC': 1626777376.0, 'UTCMS': 563}, 'DeviceName': 'NVR_Ch5'}`

**Comment by voskuh** [comment]
I also checked the output of the attach function of EventManager.cgi. The output shows (omitted the data array) 
```
Code=CrossRegionDetection;action=Start;index=2;data={
Code=VideoMotion;action=Stop;index=2;data={
Code=CrossRegionDetection;action=Stop;index=2;data={
Code=VideoMotion;action=Start;index=3;data={
Code=VideoMotion;action=Stop;index=3;data={
Code=CrossRegionDetection;action=Start;index=0;data={
Code=CrossRegionDetection;action=Stop;index=0;data={
Code=VideoMotion;action=Start;index=3;data={
Code=VideoMotion;action=Stop;index=3;data={
```
The index represents the channel (where channel 1 = index 0, channel 2 = index 1, channel 3 = index 2, etc).

**Comment by voskuh** [comment]
I'm trying to add a DVR, which has a lot of similarities with a NVR, exept for how the connection with the channels is made. See #59  for more details. 
When I add a deviceType as a workaround, there is another problem. 
```
Traceback (most recent call last):
File "/config/custom_components/dahua/__init__.py", line 240, in _async_update_data
results = await asyncio.gather(*coros)
File "/config/custom_components/dahua/client.py", line 170, in async_common_config
return await self.get(url)
File "/config/custom_components/dahua/client.py", line 574, in get
raise exception
File "/config/custom_components/dahua/client.py", line 560, in get
response.raise_for_status()
File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 1000, in raise_for_status
raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 400, message='Bad Request', url=URL('http://192.168.2.102:80/cgi-bin/configManager.cgi?action=getConfig&name=MotionDetect&action=getConfig&name=Lighting%5B1%5D%5B0%5D')
```
Calling the following api calls all responds a bad request.
```
/cgi-bin/configManager.cgi?action=getConfig&name==Lighting
/cgi-bin/configManager.cgi?action=getConfig&name==Lighting[1][0]
/cgi-bin/configManager.cgi?action=getConfig&name=MotionDetect&action=getConfig&name=Lighting[1][0]
```
Calling` /cgi-bin/configManager.cgi?action=getConfig&name=MotionDetect` results in the Motion detection settings. So only the lighting options don't work on my device. What does the lighting function exactly. Maybe there is another way to retrieve this option from the channels.

**Comment by Heire** [comment]
First of all nice work in this integration, been trying a few things with my NVR and I noticed something strange.
When adding a camera and thus choosing a channel it adds the "channel number +1".
So if I choose channel number 2, it chooses 3:
 Error fetching information from /cgi-bin/snapshot.cgi?channel=3 - 400, message='Bad Request', url=URL('http://xxx:80/cgi-bin/snapshot.cgi?channel=3')

Also this one shows errors:
http://xxx:80/cgi-bin/coaxialControlIO.cgi?action=getStatus&channel=1
And if i run it on my nvr it gives:
Error
Bad Request!
I have no idea what this is? seems more for a DVR approach then NVR?

Also can't get the video to work although the url is fine in vlc, something i need to check out.

**Comment by rroller** [comment]
I've released https://github.com/rroller/dahua/releases/tag/0.9.14 which will only increments the channel number by 1 if the channel number is 0. I don't know if this is correct as I don't have an NVR to test with it. Can you please give it a try and let me know how it goes?

**Comment by rroller** [comment]
@voskuh I'll try to put in a fix for that to just fail gracefully if your device doesn't support the lighting option. What model cams are you using?

**Comment by rroller** [comment]
@voskuh do either of these work for you?

```
/cgi-bin/configManager.cgi?action=getConfig&name=Lighting[0]
/cgi-bin/configManager.cgi?action=getConfig&name=Lighting[1]
```

**Comment by voskuh** [comment]
> @voskuh do either of these work for you?
> 
> ```
> /cgi-bin/configManager.cgi?action=getConfig&name=Lighting[0]
> /cgi-bin/configManager.cgi?action=getConfig&name=Lighting[1]
> ```
Don't work. 
```
Error
Bad Request!
```

> @voskuh I'll try to put in a fix for that to just fail gracefully if your device doesn't support the lighting option. What model cams are you using?

I've different types of HDCVI cams which are connected to the dvr. Don't know the exact models.

**Comment by rroller** [comment]
@voskuh I released 0.9.15 just now. Do you mind seeing if that has better results for you?

**Comment by rroller** [comment]
@Mikefila thanks. Can you enable debug logs and post me the event logs? I think I'm probably missing something when triggering events... I think I need to include a channel.

**Comment by rroller** [comment]
@Mikefila and others, I released 0.9.16 to address the events issues. Please give it a try. Now each motion event should be per cam as expected. 

https://github.com/rroller/dahua/releases/tag/0.9.16

**Comment by Mikefila** [comment]
Thanks for thew quick response but it is still the same, mailbox is `index 1` and front garden is `index 0`

```
2021-08-28 14:17:23 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '0', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 14:17:23 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR Mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '0', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR Mailbox'}
2021-08-28 14:17:39 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Start', 'index': '0', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 14:17:39 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR Mailbox', 'Code': 'VideoMotion', 'action': 'Start', 'index': '0', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR Mailbox'}
2021-08-28 14:18:13 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '0', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 14:18:13 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR Mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '0', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR Mailbox'}
```

**Comment by rroller** [comment]
@Mikefila weird, all the index values are `0`. I was trying to use the index to distinguish each cam's event. It looks like I could use the name?

**Comment by Mikefila** [comment]
That is because `index 0` was the only camera triggered then. After walking past several cameras I get the following. Note cameras on `index 2` and `index 3` are not defined but do exist. I'm not sure why the are all jumbled

```
2021-08-28 14:29:35 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '1', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 14:29:35 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '1', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 14:29:35 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '1', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 14:29:42 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '2', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 14:29:42 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '2', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 14:29:42 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '2', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 14:29:45 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '3', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 14:29:45 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '3', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 14:29:45 DEBUG (MainThread) [custom_components.dahua] Cam Data received: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '3', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
```

**Comment by rroller** [comment]
Thanks, this helps a lot. I'm going to try an make another fix. I'll report back once it is ready.

**Comment by rroller** [comment]
This is the actual data that comes back from the device...

```
--myboundary
Content-Type: text/plain
Content-Length:127

Code=VideoMotion;action=Stop;index=0;data={
   "Id" : [ 0 ],
   "RegionName" : [ "Region1" ],
   "SmartMotionEnable" : false
}
```

Notice it doesn't have anything to identify the device itself. That's fine for single cams, but when we are attached to an NVR I'm not sure I know how to identify which cam the event came from.

**Comment by voskuh** [comment]
> This is the actual data that comes back from the device...
> 
> ```
> --myboundary
> Content-Type: text/plain
> Content-Length:127
> 
> Code=VideoMotion;action=Stop;index=0;data={
>    "Id" : [ 0 ],
>    "RegionName" : [ "Region1" ],
>    "SmartMotionEnable" : false
> }
> ```
> 
> Notice it doesn't have anything to identify the device itself. That's fine for single cams, but when we are attached to an NVR I'm not sure I know how to identify which cam the event came from.

The index is the right way to know which cam triggered the event.

**Comment by rroller** [comment]
I've just released 0.9.17 with some extra debug logs.  @Mikefila can you paste in your event logs as you did before?

**Comment by rroller** [comment]
I think I see a problem with how I've coded this. For single cams it works great. But for NVR's I'm creating a thread per channel to listen for events. Say you have 10 cams. I'll create 10 threads connected to the NVR to listen to events. Now all 10 threads will get the same events. And that's why we are seeing this duplication. Really I need a single thread to listen for events and have it dispatch to the correct cam integration

**Comment by voskuh** [comment]
> I think I see a problem with how I've coded this. For single cams it works great. But for NVR's I'm creating a thread per channel to listen for events. Say you have 10 cams. I'll create 10 threads connected to the NVR to listen to events. Now all 10 threads will get the same events. And that's why we are seeing this duplication. Really I need a single thread to listen for events and have it dispatch to the correct cam integration

Yes, I think that's the problem.

**Comment by rroller** [comment]
I think I can put in a short term fix for now and work on the single thread approach (which is more involved) in a future update.

**Comment by rroller** [comment]
OK 0.9.18 is released with a short term fix that hopefully fixes this duplication issue.

**Comment by voskuh** [comment]
> OK 0.9.18 is released with a short term fix that hopefully fixes this duplication issue.

As far as I can see this solved the duplication issue for now.

**Comment by Mikefila** [comment]
>  I'm not sure I know how to identify which cam the event came from.

I haven't installed  .18 this is .17

The script i use now just basically transfers the event log to mqtt. `Video Motion` the only way to identify the camera is through `index` 

```
2021-08-28 16:22:31 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 1: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Start', 'index': '6', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 16:22:31 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 0: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Start', 'index': '6', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 16:22:42 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 1: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Start', 'index': '5', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 16:22:42 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 0: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Start', 'index': '5', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 16:23:04 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 1: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '6', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 16:23:04 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 0: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '6', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
2021-08-28 16:23:15 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 1: {'name': 'XVR mailbox', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '5', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR mailbox'}
2021-08-28 16:23:15 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 0: {'name': 'XVR front garden', 'Code': 'VideoMotion', 'action': 'Stop', 'index': '5', 'data': {'SmartMotionEnable': False}, 'DeviceName': 'XVR front garden'}
```

Same goes for `crossRegion` `crossLine` etc except these events can be also be named. You can have many different IVS events from a single camera and have individual triggers for each event.

```
2021-08-28 16:18:22 DEBUG (MainThread) [custom_components.dahua] Cam Data received from channel 1: {'name': 'XVR mailbox', 'Code': 'CrossRegionDetection', 'action': 'Stop', 'index': '3', 'data': {'Action': 'Appear', 'Class': 'Normal', 'CountInGroup': 1, 'DetectRegion': [[4168, 3111], [4405, 7778], [5916, 7510], [5388, 3160]], 'EventSeq': 46, 'FrameSequence': 1697394, 'GroupID': 0, 'IndexInGroup': 0, 'Mark': 0, 'Name': 'Full Screen', 'Object': {'Action': 'Appear', 'BoundingBox': [4096, 4888, 4608, 5448], 'Center': [4352, 5168], 'Confidence': 0, 'FrameSequence': 0, 'ObjectID': 116, 'ObjectType': 'Unknown', 'RelativeID': 0, 'Source': 0.0, 'Speed': 0, 'SpeedTypeInternal': 0}, 'PTS': 32293.0, 'RuleId': 1, 'Sequence': 0, 'Source': 0.0, 'Track': None, 'UTC': 1630167491, 'UTCMS': 0}, 'DeviceName': 'XVR mailbox'}
```



Where  `'Name': 'Full Screen',`  can be a unique id for all IVS events.

This is the value template I use to get events by name, I dont need to reference the camera index
```
{{ 'ON' if value_json.data.Name == 'FrontTrip' and value_json.action == 'Start' else 'OFF'}}
```


To truncate, you can name events but not motion events.

**Comment by rroller** [comment]
@Mikefila yeah it's best to use the event bus in HA to do these automations instead of the sensors I made. The sensors work for the simple case that lots of people will use but if you have a more advanced setup then you'll need to use the event fired on the bus as you correctly point out, has all the needed data.And with the latest update I believe they are deduplicated.

**Comment by Mikefila** [comment]
This is great .18 is working separating the events. I am going to add my other cameras and I'll report if there are any issues. 

I apologize, I often leave out pertinent parts but expand on details. I spent sometime sorting through the JSON from the other script, so I'm pretty familiar with the event outputs.

I though I saw you mention about using names. I wanted to explain that names can be unique but requires additional configuration by the user, so not really a viable option. 

What is the event type for the cameras?

**Comment by rroller** [comment]
@Mikefila it's in the readme (let me know if it's missing anything and I'll update).. but it looks like this...

```
platform: event
event_type: dahua_event_received
event_data:
  name: Cam13
  Code: VideoMotion
  action: Start
```

**Comment by GaryOkie** [comment]
> @Mikefila yeah it's best to use the event bus in HA to do these automations instead of the sensors I made. The sensors work for the simple case that lots of people will use but if you have a more advanced setup then you'll need to use the event fired on the bus as you correctly point out, has all the needed data.And with the latest update I believe they are deduplicated.

I was involved in a now defunct PR to add event handling to the Amcrest Integration.   The developer had a very hard time getting it approved and gave up.   One of the reasons was he added a config yaml change which was a no-no, so he went back and added config flow code, then told there were too many changes, break the PR apart.   

See https://github.com/home-assistant/core/pull/40496#discussion_r516753220

_**balloob on Nov 13, 2020**_    
>Events should not be binary sensors to begin with. All events should be … events.

>See https://developers.home-assistant.io/docs/integration_events

 > We should remove the option to define which binary sensors are added. Instead that should be controlled by the user by enabling/disabling entities in the entity registry.

In revisiting this PR I just discovered that another developer picked up where blademckain left off and was somehow able to get event handling added to the Amcrest integration.    (https://www.home-assistant.io/integrations/amcrest/#events)

 And I see some config yaml updates were added as well for tripwire detection as binary sensors!.   Whoa.   So inconsistent.  No idea how that got approved  


I mention all of this because there is a LOT of overlap between the Amcrest integration and your new Dahua code.    I started using your code because the original code owner for Amcrest bailed out, and it lacked doorbell support, tripwire, etc.     Now I see the Amcrest integration has this capability as well (plus PTZ, and new support for NVR channels).   What it still lacks is config flow support.

I'm wondering what the possibility is for collaboration between the two integrations?    It maybe boils down to which library is supported better - python_amcrest or python DahuaVTO.

**Comment by rroller** [comment]
I don't like the HA model where you have to go through very rigorous code review and long release cycles for integrations. The HACS model is more ideal IMO where we can update integrations out of band with official HA releases. I'm just doing this for fun so I'll continue to build this integration. I also don't like the Amcrest branding. Amcrest are Dahua cameras rebranded (I know it's trivial :) ) The main blocker for me working on the Amcrest integration is the inability to get code merged because of the code review process.

**Comment by GaryOkie** [comment]
I understand!   I was not suggesting dropping what you are doing here and working on the Amcrest Integration at all.    I recall seeing somwhere in your comments that the eventual roadmap for this Dahua code was to be part of core.   

My intent was 2 parts - to show the direction the code needed to take in terms of how sensors and events should be handled from a HASS developer's perspective (and review).

And to show that there is a lot of code that can be shared.    

BTW - All my cameras and NVR's are Dahua, except for the doorbells.  Not a big deal, but I put up with the Amcrest branding  and even suggested to blademckain that he fork Amcrest to be the new Dahua integration.   I love what you are doing here and am finding it works really well for my needs.  I just had not realized that the Amcrest core code had improved so much lately as well.

**Comment by tane64** [comment]
Also using Dahua NVR.
When onvif is enabled you can access any camera and stream.  Just change the channel for other cameras.   Streams are fast in the 2021.9 whit auto resize to device.  


-rtsp_transport tcp -i rtsp://user:onvifpassword@172.16.4.102:554/cam/realmonitor?channel=1&subtype=1&unicast=true&proto=Onvif

**Comment by blademckain** [comment]
> > @Mikefila yeah it's best to use the event bus in HA to do these automations instead of the sensors I made. The sensors work for the simple case that lots of people will use but if you have a more advanced setup then you'll need to use the event fired on the bus as you correctly point out, has all the needed data.And with the latest update I believe they are deduplicated.
> 
> I was involved in a now defunct PR to add event handling to the Amcrest Integration. The developer had a very hard time getting it approved and gave up. One of the reasons was he added a config yaml change which was a no-no, so he went back and added config flow code, and then was told there were too many changes at once, break the PR apart.
> 
> In this PR discussion, Balloob reviewed it as well, and had this to say about events and binary sensors...
> 
> See [home-assistant/core#40496 (comment)](https://github.com/home-assistant/core/pull/40496#discussion_r516753220)
> 
> _**balloob on Nov 13, 2020**_
> 
> > Events should not be binary sensors to begin with. All events should be … events.
> 
> > See https://developers.home-assistant.io/docs/integration_events
> 
> > We should remove the option to define which binary sensors are added. Instead that should be controlled by the user by enabling/disabling entities in the entity registry.
> 
> In revisiting this PR I just discovered that another developer picked up where blademckain left off and was somehow able to get event handling added to the Amcrest integration. (https://www.home-assistant.io/integrations/amcrest/#events)
> 
> And I see some config yaml updates were added as well for tripwire detection as binary sensors!. Whoa. So inconsistent. No idea how that got approved
> 
> I mention all of this because there is a LOT of overlap between the Amcrest integration and your new Dahua code. I started using your code because the original code owner for Amcrest bailed out, and it lacked doorbell support, tripwire, etc. Now I see the Amcrest integration has this capability as well (plus PTZ, and new support for NVR channels). What it still lacks is config flow support.
> 
> I'm wondering what the possibility is for collaboration between the two integrations? It maybe boils down to which library is supported better - python_amcrest or python DahuaVTO.

the dissident programmer is here :)

@rroller your work is great
thanks a lot

**Comment by tomlut** [comment]
My DHI-NVR5208-8P-4KS2E with 7 cameras set up no issues. I'm getting camera feeds and intelligent motion alerts. However it takes 15 times longer than the Onvif integration to set up (60 sec as opposed to 4 sec for Onvif). 

Is this normal?

Can anything be done?

**Comment by Mikefila** [comment]
Nvr/dvr/xvr support was added after the fact. So it treats each instance as if it was a standalone ip camera. Instead of logging in once to the nvr, it logs in for each camera and slows the unit down. 

I don't think anything can be done on our side. I have 11 cameras on a ssd pi and it takes longer than a minute to load. I have 5 streams enabled.

**Comment by tomlut** [comment]
Can the logins be done simultaneously?

**Comment by zios0007** [comment]
Worth a status check after four years, because two of the things in this thread were true when they were written and are not now.

**"it logs in for each camera instead of logging in once to the NVR"** (@Mikefila, 2022). Reads are shared per host now rather than per entry. A config read is held for 300 seconds and a status read for 5, so ten channels on one recorder no longer produce ten of every request. Digest state is shared per host too, so the channels are not each doing their own challenge.

**"add one device (NVR) and all info about cameras would be detected automatically"**, which is what this issue actually asked for. #746 is open for it. Adding one channel reads the recorder's `RemoteDevice` table, offers the other channels that have a live camera behind them, and creates an entry for each one ticked. Measured on a 16 slot recorder it offered 10 of 11 candidates, named from the recorder.

It will not detect them with no input at all, because credential validation on these devices is host level and says nothing about channels, so a slot has to be probed before it is worth offering.

@rroller this looks closeable once #746 merges, unless somebody has an NVR gap that is neither of the above.
